### DISCUSSION TITLE
Technical Help Wanted

### DISCUSSION THREAD

**Discussion by PRProd** [PRProd/HA-Firemote#115] [General]
Do you have experience working with Android devices?  Are you an ADB command wizard?  Have you found a super useful command that you can trigger through Home Assistant's "Android Debug Bridge" or "Android TV Remote" integrations that you haven't seen anywhere else?  Then this is the place for you!

Below, you will find some sub-thread topics for various commands and/or un-cracked puzzles that's I'd love to solve.  If you can help, please reply INSIDE the appropriate thread.  Thanks!

**Comment by PRProd** [comment]
# ADB Command to reconnect a physical FireTV remote
## What
 I'm looking for an ADB command that could be used to re-connect a paired, but disconnected FireTV remote back to its FireTV device.

## Why
Some Firemote button presses work because they are masquerading as commands coming from a physical remote.  When a Firestick is restarted, the communication channel that a physical remote uses is severed until a button on the physical remote is pressed which restores the connection.  When the channel is severed, Firemote cannot re-open the communication, or send commands through that channel until a button press on the physical remote occurs and re-opens the communication channel.  This frustration is noted in the [FAQ](https://github.com/PRProd/HA-Firemote#why-dont-any-of-the-buttons-on-the-firemote-work-at-all).

## Technical Details
When external devices are connected, an input path at /dev/input/event{NUMBER} is created.  On device power off or restart, these input paths are removed.  For example, if a Firemote user presses the "Down" button on the DPad, the following command might be sent
```bash
adb shell sendevent /dev/input/event5 1 108 1 && sendevent /dev/input/event5 0 0 0 && sendevent /dev/input/event5 1 108 0 && sendevent /dev/input/event5 0 0 0
```
If "event5" doesn't exist because a physical button on the remote has not yet been pressed, then this command will fail with the error message "sendevent: /dev/input/event5: No such file or directory"

## What hasn't worked
I've tried the following commands, but they have not produced any productive results:
 - `adb shell am broadcast -a com.amazon.firetv.action.RECONNECT_REMOTE`
 - `adb shell am broadcast -a com.amazon.firetv.remote.action.RECONNECT_REMOTE`
 - `adb shell am broadcast -a com.amazon.firetv.intent.action.PAIR_REMOTE`
 - `adb shell am broadcast -a android.intent.action.FIRETV_REMOTE_CONTROL_RECONNECT`
 - `adb shell am broadcast -a android.intent.action.FIRETV_REMOTE_PAIRING`
 - `adb shell am broadcast -a android.intent.action.FIRETV_REMOTE_PAIR`
 - `adb shell am broadcast -a android.intent.action.RECONNECT_BT_DEVICE`
 - `adb shell am broadcast -a android.intent.action.RECONNECT_BT_DEVICE --es name "Amazon Fire TV Remote"`
 - `adb shell am broadcast -a android.intent.action.TV_REMOTE_CONTROL_RECONNECT`
 - `adb shell am broadcast -a android.intent.action.TV_REMOTE_PAIRING`
 - `adb shell am broadcast -a android.intent.action.ACTION_RECONNECT_DEVICE`
 - `adb shell am broadcast -a android.intent.action.ACTION_REMOTE_DEVICE_PAIR`
 - `adb shell am broadcast -a android.intent.action.ACTION_REMOTE_DEVICE_CONNECTED --es name "Amazon Fire TV Remote"`
 - `adb shell am broadcast -a android.intent.action.ACTION_TV_REMOTE_DEVICE_CONNECTED -d android.provider.Settings.ACTION_TV_REMOTE_DEVICE_CONNECTED`
 - `adb shell am broadcast -a android.intent.action.RECONNECT_REMOTE_CONTROL`
 - `adb shell am broadcast -a android.intent.action.RECONNECT_INPUT_DEVICE`
 - `adb shell am broadcast -a android.intent.action.RECONNECT_HDMI_CEC`
 - `adb shell am broadcast -a android.intent.action.HDMI_CEC_CONNECT_REMOTE`
 - `adb shell am broadcast -a android.intent.action.DEVICE_PAIRED`
 - `adb shell am broadcast -a android.intent.action.DEVICE_PAIRING_REQUEST`
 - `adb shell input keyevent 133` (suggested by AI chatbot)
 - `adb shell input keyevent 134`  (suggested by AI chatbot)
 - `adb shell input keyevent 5`  (suggested by AI chatbot)
 - `adb shell input keyevent KEYCODE_BACK` (suggested by AI chatbot)

**Reply by reinuke** [reply]
Since remotes are BLE devices it means they aren't "always online" but rather "always offline (until a key press)" . You could try 
adb shell input keyevent KEYCODE_PAIRING
To see if it the remote is able to see the request.. 
I doubt for some reason.. 

I downloaded Android Studio and started browsing my MiBox for fun :D 
Found a file named KeyEvent.java that lists all of the accepter keypresses

**Reply by PRProd** [reply]
Thanks for the guess, but nope... that won't do it

**Comment by PRProd** [comment]
# ADB Command for Alexa Voice Remote Pro "Find My Remote"
## What
 I'm looking for an ADB command that will initiate the remote finder feature for a connected [Alexa Voice Remote Pro](https://www.amazon.com/Introducing-alexa-voice-remote-pro/dp/B09RX4HKTD/ref=sr_1_1).

## Why
Discussions show that this feature in Firemote has been very handy for people who have used it with their NVIDIA Shield devices.  The Amazon pro remote is able to "ring" like the Shield remote, but I've been unsuccessful in being able to trigger the ringing without using the FireTV mobile app and/or speaking to Alexa.  I'd love to be able to make the "Find My Remote" feature available for FireTV users who have the pro remote.

## What hasn't worked
I've tried the following commands, but they have not produced any productive results:
 - `adb shell am broadcast -a com.amazon.firetv.remotefinder.ACTION_FIND_REMOTE`
 - `adb shell am broadcast -a com.amazon.firetv.remotefinder.action.START_FIND_REMOTE`
 - `adb shell am broadcast -a com.amazon.firetv.remotefinder.action.FIND_REMOTE`
 - `adb shell am broadcast -a com.amazon.firetv.remotefinder.action.FIND_MY_REMOTE`
 - `adb shell am broadcast -a com.amazon.firetv.remotefinder.FIND_REMOTE`
 - `adb shell am broadcast -a com.amazon.firetv.remotefinder.LAUNCH`
 - `adb shell am broadcast -a com.amazon.firetv.remotefinder.SHOW_REMOTE_FINDER`
 - `adb shell am broadcast -a com.amazon.firetv.remotefinder.REQUEST_REMOTE_FINDER`
 - `adb shell am broadcast -a com.amazon.firetv.remotecontrol.action.FIND_REMOTE`
 - `adb shell am broadcast -a com.amazon.firetv.remote.action.FIND_REMOTE`
 - `adb shell am broadcast -a com.amazon.firetv.action.LAUNCH_FIND_REMOTE`
 - `adb shell am broadcast -a com.amazon.firetv.action.LAUNCH_FIND_MY_REMOTE`
 - `adb shell am broadcast -a com.amazon.firetv.action.FIND_REMOTE`
 - `adb shell am broadcast -a com.amazon.firetv.action.LOCATE_REMOTE`
 - `adb shell am broadcast -a com.amazon.firetv.action.REMOTE_FIND`
 - `adb shell am broadcast -a com.amazon.firetv.alexa.action.FIND_REMOTE`
 - `adb shell am broadcast -a com.amazon.firetv.remote.action.LOCATE_REMOTE`
 - `adb shell am broadcast -a com.amazon.tv.remotefinder.ACTION_FIND_REMOTE`
 - `adb shell am broadcast -a com.amazon.tv.remote.action.LOCATE_REMOTE`
 - `adb shell am broadcast -a com.amazon.tv.remotecontrol.action.FIND_REMOTE`
 - `adb shell am broadcast -a com.amazon.tv.action.ALEXA_FIND_MY_REMOTE`
 - `adb shell am broadcast -a com.amazon.firetv.intent.action.FIND_REMOTE`
 - `adb shell am broadcast -a com.amazon.intent.action.LAUNCH_FIND_REMOTE`
 - `adb shell am broadcast -a com.amazon.intent.action.REMOTE_FIND`
 - `adb shell am broadcast -a com.amazon.alexa.action.FIND_MY_REMOTE`
 - `adb shell am broadcast -a com.amazon.alexa.action.FIND_REMOTE`
 - `adb shell am broadcast -a com.amazon.alexa.action.SAY_PLAIN_TEXT --es text "Alexa, Find My Remote"`
 - `adb shell am broadcast -a com.amazon.alexa.action.VOICE_COMMAND -d "alexa:find_my_remote"`
 - `adb shell am broadcast -a com.amazon.alexa.action.VOICE_COMMAND -d "alexa:findmyremote"`
 - `adb shell am broadcast -a com.amazon.alexa.action.PLAY_SOUND -d "alexa:findmyremote"`
 - `adb shell am broadcast -a com.amazon.alexa.action.INVOKE_SKILL -n com.amazon.dee.skill.findmyremote`
 - `adb shell am start -a com.amazon.firetv.remotefinder.action.START_FIND_REMOTE`
 - `adb shell am start -a com.amazon.dee.intent.action.VOICE_COMMAND -d "Alexa, Find My Remote"`
 - `adb shell am start -a com.amazon.alexa.action.FIND_REMOTE`
 - `adb shell am start -a com.amazon.alexa.action.VOICE_SEARCH -d "Alexa, Find My Remote"`
 - `adb shell am start -a com.amazon.alexa.action.SAY_SIMPLE_COMMAND -d "Alexa, Find My Remote"`
 - `adb shell am start -a com.amazon.alexa.action.SAY_PLAIN_TEXT -d "Alexa, Find My Remote"`
 - `adb shell am start -a com.amazon.alexa.action.QUERY -e action "FindMyRemote"`
 - `adb shell am start -a com.amazon.dee.intent.action.LAUNCH_APP -n com.amazon.dee.app -e action "alexa_remote_find"`
 - `adb shell am start -a android.intent.action.MAIN -n com.amazon.dee.app/com.amazon.dee.app.VoiceActivity -e alexa_command "Find My Remote"`
 - `adb shell am start -a android.intent.action.VOICE_COMMAND -d "Alexa, Find My Remote"`
 - `adb shell am broadcast -a com.amazon.alexa.action.INVOKE_CUSTOM_COMMAND -d com.amazon.alexa.extra.CUSTOM_COMMAND_NAME="FindMyRemote"`
 - `adb shell input keyevent 67`
 - `adb shell input keyevent 85`
 - `adb shell input keyevent 86`
 - `adb shell input keyevent 87`
 - `adb shell input keyevent 89`
 - `adb shell input keyevent 102`
 - `adb shell input keyevent 105`
 - `adb shell input keyevent 115`
 - `adb shell input keyevent 126`
 - `adb shell input keyevent 130`
 - `adb shell input keyevent 136`
 - `adb shell input keyevent 139`
 - `adb shell input keyevent 167`
 - `adb shell input keyevent 182`
 - `adb shell input keyevent 187`
 - `adb shell input keyevent 127`
 - `adb shell input keyevent 130`
 - `adb shell input keyevent 224`
 - `adb shell input keyevent 300`

**Comment by PRProd** [comment]
# Close NVIDIA Shield Settings on 2nd button press
## What
 I'm looking for an ADB command that could be used to close the settings menu if the button is pressed while menu is open

## Why
On the physical NVIDIA remote, when the settings (hamburger) menu button is pressed, the menu opens.  When it is pressed a second time, the settings menu closes.  Currently, Firemote is unable to close this open menu, so clicking the button again just keeps the menu open.  This is inconvenient, and was also brought up by a user in #198

## Technical Details
According to this [reddit thread](https://www.reddit.com/r/ShieldAndroidTV/comments/ih81m1/adb_and_keyevent_for_menu_button/jcibw0q/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button), there was a "technical" reason why this button press couldn't be simulated in the same way a regular remote button press is.  The advice given in the thread is ultimately the way that Firemote launches this menu.

## What hasn't worked
I've tried the following commands, but they have not produced any productive results:
 - `adb shell am start -a android.settings.SETTINGS`
